src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java

```diff
@@ -1,37 +1,91 @@
 package com.es.phoneshop.web;
 
-import com.es.phoneshop.model.product.ArrayListProductDao;
-import com.es.phoneshop.model.product.NullValuePassedException;
-import com.es.phoneshop.model.product.ProductDao;
+import com.es.phoneshop.dao.ArrayListProductDao;
+import com.es.phoneshop.exception.NullValuePassedException;
+import com.es.phoneshop.dao.ProductDao;
+import com.es.phoneshop.exception.OutOfStockException;
+import com.es.phoneshop.model.recentlyviewed.RecentlyViewedProducts;
+import com.es.phoneshop.model.cart.Cart;
+import com.es.phoneshop.service.CartService;
+import com.es.phoneshop.service.HttpSessionCartService;
+import com.es.phoneshop.service.HttpSessionRecentlyViewedProductsService;
+import com.es.phoneshop.service.RecentlyViewedProductsService;
 
 import javax.servlet.ServletConfig;
 import javax.servlet.ServletException;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.io.IOException;
+import java.text.NumberFormat;
+import java.text.ParseException;
 
 public class ProductDetailsPageServlet extends HttpServlet {
     private ProductDao productDao;
+    private CartService cartService;
+    private RecentlyViewedProductsService recentlyViewedProductsService;
 
     @Override
     public void init(ServletConfig config) throws ServletException {
         super.init(config);
         productDao = ArrayListProductDao.getInstance();
+        cartService = HttpSessionCartService.getInstance();
+        recentlyViewedProductsService = HttpSessionRecentlyViewedProductsService.getInstance();
     }
 
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-        String productId = request.getPathInfo();
+        String productIdString = request.getPathInfo();
+        Long productId;
+        RecentlyViewedProducts products = recentlyViewedProductsService.getRecentlyViewedProducts(request);
         try {
-            request.setAttribute("product", productDao.getProduct(Long.valueOf(productId.substring(1))));
+            productId = Long.valueOf(productIdString.substring(1));
+            request.setAttribute("product", productDao.getProduct(productId));
+            request.setAttribute("cart", cartService.getCart(request));
+            request.setAttribute("recentlyViewed", products);
         } catch (NullValuePassedException e) {
             throw new RuntimeException("Null value passed", e);
         }
+        recentlyViewedProductsService.add(products, productId);
```

Comment:
1) When i open product for the first time, it's already displayed as viewed before. This can be confusing for user
2) I think that products objects can be quite large to store them as session attributes